indico/modules/rb_new/client/js/components/RoomListFilters.jsx

```diff
@@ -0,0 +1,182 @@
+/* This file is part of Indico.
+ * Copyright (C) 2002 - 2018 European Organization for Nuclear Research (CERN).
+ *
+ * Indico is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Indico is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import propTypes from 'prop-types';
+import React from 'react';
+import {Button, Form, Icon, Input, Popup, Select} from 'semantic-ui-react';
+import {DebounceInput} from 'react-debounce-input';
+
+import {Translate} from 'indico/react/i18n';
+import {parseRoomListFiltersText} from '../util';
+
+import './RoomListFilters.module.scss';
+
+
+export default class RoomListFilters extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            filtersVisible: false,
+            text: '',
+            building: '',
+            floor: ''
+        };
+    }
+
+    componentDidMount() {
+        const {fetchBuildings} = this.props;
+        fetchBuildings();
+    }
+
+    handleVisibleChange(visible) {
+        this.setState({filtersVisible: visible});
+    }
+
+    applyFilters() {
+        const {onConfirm} = this.props;
+        this.setState({filtersVisible: false});
+        onConfirm();
+    }
+
+    handleFiltersChange(filter, value) {
+        const {setTextParamFilter, setAdvancedParamFilter, filters: {text}} = this.props;
+        const stateUpdates = {[filter]: value};
+
+        if (filter === 'building') {
+            stateUpdates.floor = '';
+        } else if (filter === 'text') {
+            const parsedValues = parseRoomListFiltersText(value);
+
+            stateUpdates.building = parsedValues.building;
+            stateUpdates.floor = parsedValues.floor;
+        }
+
+        this.setState(stateUpdates, () => {
+            let textValue = '';
+            const {building, floor} = this.state;
+
+            if (filter === 'text') {
+                textValue = value;
+            } else {
+                if (building) {
+                    textValue = `building:${building}`;
+                }
+
+                if (floor) {
+                    textValue += ` floor:${floor}`;
```

Comment:
let's build an array and `.join(' ')` it later?

Reply:
also, what about a mapping for the criteria and iterating over that? something similar to this maybe:

```javascript
const stateToKeys = {building: 'building', floor: floor};
const searchParts = Object.entries(stateToKeys).map(([stateKey, searchKey]) => {
    // TODO: skip entries with no value (maybe a `filter()` before the `map()`
    return `${searchKey}:${this.state[stateKey]}`;
});
searchParts.push(text);  // the remaining searc text that's not a `foo:aar` option
seachParts.join(' ');
```